AppCenterCrashes/AppCenterCrashes/Internals/Util/MSCrashesCategory.m

```diff
@@ -0,0 +1,210 @@
+ *        PLCrashReporter to stop all running threads every time an exception occurs even if will be handled right
+ *        away, e.g. by a system framework.
+ *
+ *
+ * Solution B:
+ *
+ *   Overwrite and extend specific methods of `NSApplication`. Can be implemented via subclassing NSApplication or
+ *   by using a category.
+ *
+ *   Benefits:
+ *
+ *     1. Solves scenarios 2 (by overwriting `reportException:`) and 3 (by overwriting `sendEvent:`)
+ *
+ *     2. Subclassing approach isn't enforcing the mechanism onto apps and let developers opt-in.
+ *        (Category approach would enforce it and rather be a problem of this soltuion.)
+ *
+ *     3. Safe. Doesn't use private API.
+ *
+ *  Problems:
+ *
+ *     1. Does not automatically solve scenario 1. Developer would have to put all that code into @try @catch blocks
+ *
+ *     2. Not a clean implementation, rather feels like a workaround.
+ *
+ *     3. Not clear if this solves all scenarios there can be.
+ *
+ *
+ * References:
+ *   https://developer.apple.com/library/mac/documentation/cocoa/Conceptual/Exceptions/Tasks/ControllingAppResponse.html#//apple_ref/doc/uid/20000473-BBCHGJIJ
+ *   http://stackoverflow.com/a/4199717/474794
+ *   http://stackoverflow.com/a/3419073/474794
+ *   http://macdevcenter.com/pub/a/mac/2007/07/31/understanding-exceptions-and-handlers-in-cocoa.html
+ *
+ */
+@implementation NSApplication (CrashException)
+
++ (void)swizzleReportException {
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    Class class = [self class];
+    Method originalMethod = class_getInstanceMethod(class, @selector(reportException:));
+    IMP swizzledImp = class_getMethodImplementation(class, @selector(ms_reportException:));
+    reportExceptionOriginalImp = method_setImplementation(originalMethod, swizzledImp);
+  });
+}
+
++ (void)swizzleSendEvent {
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    Class class = [self class];
+    Method originalMethod = class_getInstanceMethod(class, @selector(reportException:));
+    IMP swizzledImp = class_getMethodImplementation(class, @selector(ms_reportException:));
+    sendEventOriginalImp = method_setImplementation(originalMethod, swizzledImp);
+  });
+}
+
+/*
+ * Solution for Scenario 2
+ *
+ * Catch all exceptions that are being logged to the console and forward them to our
+ * custom UncaughtExceptionHandler
+ */
+- (void)ms_reportException:(NSException *)exception {
+
+  // Don't invoke the registered UncaughtExceptionHandler if we are currently debugging this app!
+  if (![MSAppCenter isDebuggerAttached] && exception) {
+
+    /*
+     * We forward this exception to PLCrashReporters UncaughtExceptionHandler.
+     * If the developer has implemented their own exception handler and that one is invoked before PLCrashReporters exception handler and
+     * the developers exception handler is invoking this method it will not finish it's tasks after this call but directly jump into
```

Comment:
```suggestion
     * the developers exception handler is invoking this method it will not finish its tasks after this call but directly jump into
```